### ISSUE TITLE
Error in UseMethod("model_performance") :    no applicable method for 'model_performance' applied to an object of class "prediction"

### ISSUE THREAD

**Issue by fuehrding** [ipa-tys/ROCR#6]
Good morning, I use ROCR for the first time, the predict and the prediction functions worked fine so far, but I run into trouble using performance, which I suppose is weird. I started a clean project in order to make sure that no other package would interfere, but I always get the same result (see: below). I consulted other communication platforms like the StackExchange whether someone else experienced this, but couldn't find a hint or solution.
Any idea would be appreciated what could cause this error.

Thanks a lot in advance

###
install.packages("ROCR")
library(ROCR)
###
class(pr)
[1] "prediction"
attr(,"package")
[1] "ROCR"
### 
prf = performance(pr, measure = "tpr", x.measure = "fpr")
Error in UseMethod("model_performance") : 
  no applicable method for 'model_performance' applied to an object of class "prediction"
###
performance(pr,"tpr","fpr")
Error in UseMethod("model_performance") : 
  no applicable method for 'model_performance' applied to an object of class "prediction"

**Comment by FelixErnst** [comment]
Hi,

can you please provide a reproducible example and the output of `sessionInfo()`? 
This includes also example data to create the object `pr`, which is not stated in your example code.

As a control please run the code snippet and let me know about the output:
```
data(ROCR.simple)
pred <- prediction( ROCR.simple$predictions, ROCR.simple$labels)
prf <- performance(pred, measure = "tpr", x.measure = "fpr")
```

From what I can tell, the problem might lie outside of the ROCR package, since no `UseMethod` statemend is used in ROCR package. Also the term `model` is not used within ROCR.

Felix

**Comment by fuehrding** [comment]
Hi, thanks a lot for the swift response, I hope what I am sending now is sufficient, otherwise it will be a long piece of code and I won't take too much of your time.

Thanks, again, for your support, kind regards, Petra

### First, my R output to your code input
> data(ROCR.simple)
> pred <- prediction( ROCR.simple$predictions, ROCR.simple$labels)
> prf <- performance(pred, measure = "tpr", x.measure = "fpr")
Error in UseMethod("model_performance") : 
  no applicable method for 'model_performance' applied to an object of class "prediction"

### Second, the sessionInfo

> sessionInfo()
R version 3.6.1 (2019-07-05)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18362)

Matrix products: default

locale:
[1] LC_COLLATE=German_Germany.1252  LC_CTYPE=German_Germany.1252    LC_MONETARY=German_Germany.1252
[4] LC_NUMERIC=C                    LC_TIME=German_Germany.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] performance_0.4.5.1 ROCR_1.0-7          gplots_3.0.3        BIEN_1.2.4          RPostgreSQL_0.6-2  
 [6] DBI_1.0.0           raster_3.0-12       maptools_0.9-9      rgeos_0.5-2         rgdal_1.4-4        
[11] sp_1.4-0            plyr_1.8.5         

loaded via a namespace (and not attached):
 [1] gtools_3.8.1       remotes_2.1.1.9000 sf_0.8-0           lattice_0.20-38    testthat_2.3.2    
 [6] usethis_1.6.1      rlang_0.4.5        pkgbuild_1.0.7     e1071_1.7-2        foreign_0.8-71    
[11] glue_1.4.0         withr_2.1.2        sessioninfo_1.1.1  foreach_1.4.7      bayestestR_0.6.0  
[16] caTools_1.18.0     devtools_2.3.0     codetools_0.2-16   memoise_1.1.0      callr_3.4.3       
[21] doParallel_1.0.15  ps_1.3.0           curl_4.0           parallel_3.6.1     class_7.3-15      
[26] fansi_0.4.1        Rcpp_1.0.3         KernSmooth_2.23-15 backports_1.1.4    classInt_0.4-2    
[31] gdata_2.18.0       fasterize_1.0.2    desc_1.2.0         pkgload_1.0.2      fs_1.3.1          
[36] digest_0.6.25      insight_0.8.3      processx_3.4.1     grid_3.6.1         rprojroot_1.3-2   
[41] bitops_1.0-6       cli_2.0.2          tools_3.6.1        magrittr_1.5       crayon_1.3.4      
[46] ape_5.3            ellipsis_0.3.0     prettyunits_1.0.2  assertthat_0.2.1   rstudioapi_0.11   
[51] iterators_1.0.12   R6_2.4.1           units_0.6-3        nlme_3.1-140       compiler_3.6.1   

### Third, my file: pr to use as input for the predictions, it has been developed and I have exported and saved it as an Excel file, I hope it will work
[pr.xlsx](https://github.com/ipa-tys/ROCR/files/4563606/pr.xlsx)

### Fourth, information regarding used input data fro ROCR, you find in below calls ("Fifth")
> logreg_mau_fle

Call:  glm(formula = present ~ elev + bio1 + bio12, family = "binomial", 
    data = mau_fle_abs@data)

Coefficients:
(Intercept)         elev         bio1        bio12  
 -9.451e+00    6.645e-05    2.730e-02    1.321e-03  

Degrees of Freedom: 10370 Total (i.e. Null);  10367 Residual
  (5 observations deleted due to missingness)
Null Deviance:	    14380 
Residual Deviance: 11280 	AIC: 11290

> mau_fle_abs@data
        longitude    latitude present elev bio1 bio12
1858    -76.25066   1.2500000       0 1331  192  4198
1556    -63.25066   4.2500000       0  747  223  2963
3334    -36.25066  -7.7500000       0  529  240   481
340     -97.75066  29.2500000       0  107  212   845
2829    -63.25066  -5.2500000       0   52  257  2456
2921    -62.75066  -5.7500000       0   37  258  2168
4904    -62.75066 -18.7500000       0  367  245   976
2236    -78.75066  -1.7500000       0 3304   95   665
3300    -53.25066  -7.7500000       0  293  248  1897
1028    -83.78541  10.7273484       0   25  263  4504
3773    -74.75066 -10.7500000       0  810  228  2041
5326    -58.25066 -22.7500000       0   91  244  1220
1692    -74.25066   2.7500000       0  365  247  1969
2406    -66.25066  -2.7500000       0   83  253  2580 
...
974     -85.75066  13.7500000       0  425  242  1329
539     -99.75066  23.2500000       0 1161  199   386
982     -86.25066  13.2500000       0 1322  199  1499
 [ reached 'max' / getOption("max.print") -- omitted 10210 rows ]

### Fifth: my code, please mind that the two steps before the issue occurred are without any errors, hence I set it now in comments

### p = predict(logreg_mau_fle, newdata=mau_fle_abs@data, type="response")  ### ok
### pr = prediction(p, mau_fle_abs@data$present) ### ROCR: both parameters must have same dimensions, ok, both have 10,376 entries
prf = performance(pr, measure = "tpr", x.measure = "fpr") ### ROCR, not ok

**Comment by FelixErnst** [comment]
Hi Petra,

Your input works, which suggest that the input in your session is not exactly as saved in the file.
I used R 4.0.0, which was release last week, but that should not make a huge difference, since `ROCR` is stable.

```
> pr_input <- read.csv("pr.csv", sep=";", fileEncoding = "UTF-8-BOM")
> head(pr_input)
  predictions labels
1  0.80628749      0
2  0.64601946      0
3  0.09721202      0
4  0.07309895      0
5  0.69288383      0
6  0.61286402      0
> pred <- prediction(pr_input$predictions, pr_input$labels)
> perf <- performance(pred, measure = "tpr", x.measure = "fpr")
> plot(perf)
```

![image](https://user-images.githubusercontent.com/17824332/80802170-e1f93f00-8bae-11ea-897f-27c644889dfa.png)

So apart from running this test, I cannot help you to solve your problem, since `ROCR` works fine. I would suggest that you make sure, that `p` is actually a list and not something else (e.g. a model, which return `is.list() == TRUE`). Maybe you want to run `str(p)` to make sure. `predict` can output many types, since it depends on the model used.

edit: Since my code snippet also did throw an error, maybe updating packages can solve your problem as well. You should also check, whether your R installation is maybe at fault or your R session has unintended consequences for `ROCR`.

Felix

<details>
<summary>sessionInfo</summary>
<pre>
> sessionInfo()
R version 4.0.0 (2020-04-24)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default

locale:
[1] LC_COLLATE=German_Germany.1252  LC_CTYPE=German_Germany.1252    LC_MONETARY=German_Germany.1252 LC_NUMERIC=C                   
[5] LC_TIME=German_Germany.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ROCR_1.0-7   gplots_3.0.3

loaded via a namespace (and not attached):
[1] compiler_4.0.0     tools_4.0.0        KernSmooth_2.23-16 gdata_2.18.0       caTools_1.18.0     bitops_1.0-6      
[7] gtools_3.8.2      
</pre>
</details>

**Comment by fuehrding** [comment]
Thanks Felix, have a lovely weekend

**Comment by FelixErnst** [comment]
Thank you as well. Let me know, if you solve the problem.

I will close the issue for now, but feel free to reopen it, if you have additional questions.

**Comment by fuehrding** [comment]
Hi Felix,

I have now "switched off" all active packages with the exception of ROCR
and it works now, so: apparently, at least on my PC, a package interferes
with ROCR and prevents orderly execution.

I only wanted to let you know, just in case someone else has a similar
issue.

Cheers, have a nice day and weekend, Petra

Freundliche Grüsse / Kind Regards

Petra Führding-Potschkat
mailto: fuehrding@googlemail.com
cell (DE): +49 171 3054 762
cell (ZA): +27 712 84 1126
XING:
https://www.xing.com/app/profile?op=activitystream;name=PetraElisabeth_FuehrdingPotschkat



Am Fr., 1. Mai 2020 um 13:59 Uhr schrieb Felix Ernst <
notifications@github.com>:

**Comment by FelixErnst** [comment]
Thank you!

Could you narrow it down to the package causing the issue?

**Comment by fuehrding** [comment]
Hi Felix,

could you please remove my private data from the thread? I don't want to
get unpleasant phone calls or hackers or spam emails or visits on my public
profile.

Thank you, kind regards, Petra


Am Fr., 1. Mai 2020 um 14:40 Uhr schrieb Felix Ernst <
notifications@github.com>:

**Comment by fuehrding** [comment]
Hi Felix, I had to delete my email entries due to data privacy issues. Let me quickly summarize here what I wrote yesterday:

Hi Felix,

I have now "switched off" all active packages with the exception of ROCR and it works now, so: apparently, at least on my PC, a package interferes with ROCR and prevents orderly execution.

I only wanted to let you know, just in case, someone else has a similar issue.

Cheers, have a nice day and weekend, Petra

02/05/2020

The package which conflicted with ROCR was the performance package, but I don't know how exactly the two interfered with my PC, so sorry.

Cheers, Petra
